lib/web/apiserver.go

```diff
@@ -289,16 +293,149 @@ func NewHandler(cfg Config, opts ...HandlerOption) (*APIHandler, error) {
 		return nil, trace.Wrap(err)
 	}
 
+	if cfg.MinimalReverseTunnelRoutesOnly {
+		h.bindMinimalEndpoints()
+	} else {
+		h.bindDefaultEndpoints(challengeLimiter)
+	}
+
+	// if Web UI is enabled, check the assets dir:
+	var indexPage *template.Template
+	if cfg.StaticFS != nil {
+		index, err := cfg.StaticFS.Open("/index.html")
+		if err != nil {
+			h.log.WithError(err).Error("Failed to open index file.")
+			return nil, trace.Wrap(err)
+		}
+		defer index.Close()
+		indexContent, err := io.ReadAll(index)
+		if err != nil {
+			return nil, trace.ConvertSystemError(err)
+		}
+		indexPage, err = template.New("index").Parse(string(indexContent))
+		if err != nil {
+			return nil, trace.BadParameter("failed parsing index.html template: %v", err)
+		}
+
+		h.Handle("GET", "/web/config.js", httplib.MakeHandler(h.getWebConfig))
+	}
+
+	routingHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// request is going to the API?
+		if strings.HasPrefix(r.URL.Path, apiPrefix) {
+			http.StripPrefix(apiPrefix, h).ServeHTTP(w, r)
+			return
+		}
+
+		// request is going to the web UI
+		if cfg.StaticFS == nil {
+			w.WriteHeader(http.StatusNotImplemented)
+			return
+		}
+
+		// redirect to "/web" when someone hits "/"
+		if r.URL.Path == "/" {
+			http.Redirect(w, r, "/web", http.StatusFound)
+			return
+		}
+
+		// serve Web UI:
+		if strings.HasPrefix(r.URL.Path, "/web/app") {
+			httplib.SetStaticFileHeaders(w.Header())
+			http.StripPrefix("/web", makeGzipHandler(http.FileServer(cfg.StaticFS))).ServeHTTP(w, r)
+		} else if strings.HasPrefix(r.URL.Path, "/web/") || r.URL.Path == "/web" {
+			csrfToken, err := csrf.AddCSRFProtection(w, r)
+			if err != nil {
+				h.log.WithError(err).Warn("Failed to generate CSRF token.")
+			}
+
+			session := struct {
+				Session string
+				XCSRF   string
+			}{
+				XCSRF: csrfToken,
+			}
+
+			ctx, err := h.AuthenticateRequest(w, r, false)
+			if err == nil {
+				resp, err := newSessionResponse(ctx)
+				if err == nil {
+					out, err := json.Marshal(resp)
+					if err == nil {
+						session.Session = base64.StdEncoding.EncodeToString(out)
+					}
+				} else {
+					h.log.WithError(err).Debug("Could not authenticate.")
+				}
+			} else {
+				h.log.WithError(err).Debug("Could not authenticate.")
```

Comment:
The nesting in this entire function is making me pull me hair out. Since, I assume, this is mostly moving legacy code I think we can leave as-is, but pulling bits of the code out into separate functions and doing early returns on errors would make wonders here.